### PR TITLE
Refactor vars to avoid test failures in storage/remote with -count > 1

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

Fixes: #7642

### CODE REVIEW

**Comment by Harkishen-Singh** [comment]
cc @roidelapluie @cstyan

File: storage/remote/queue_manager.go [hunk 1]
```diff
@@ -57,6 +57,7 @@ type queueManagerMetrics struct {
 	enqueueRetriesTotal   prometheus.Counter
 	sentBatchDuration     prometheus.Histogram
 	highestSentTimestamp  *maxGauge
+	highestRecvTimestamp  *maxGauge
```
Comment (cstyan) [inline storage/remote/queue_manager.go:60]: this metric should not be a part of the queue manager metrics, we should just pass a reference to the metric from write storage when we create the queue manager

File: storage/remote/write.go [hunk 2]
```diff
@@ -74,6 +69,14 @@ func NewWriteStorage(logger log.Logger, reg prometheus.Registerer, walDir string
 		flushDeadline:     flushDeadline,
 		samplesIn:         newEWMARate(ewmaWeight, shardUpdateDuration),
 		walDir:            walDir,
+		highestTimestamp: &maxGauge{
+			Gauge: prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (cstyan) [inline storage/remote/write.go:73]: since you're not using promauto here you need to register this metric

**Review by cstyan — CHANGES_REQUESTED** [review]
Thanks for taking this on. Looking good but some changes are needed.

File: storage/remote/write.go (hunk 2 above)
Comment (Harkishen-Singh) [reply storage/remote/write.go:73]: With the current setup, I guess that caused panics due to duplicate registering. Hence, did a careful placement of this.

Comment (Harkishen-Singh) [reply storage/remote/write.go:73]: Oh yes. That got missed in a mistake. Thanks for pointing it out.

File: storage/remote/queue_manager.go (hunk 1 above)
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go]: Done. Hence, resolving.

**Comment by Harkishen-Singh** [comment]
Addressed the issues @cstyan.

File: storage/remote/intern_test.go [hunk 3]
```diff
@@ -36,52 +37,55 @@ func TestIntern(t *testing.T) {
 }
 
 func TestIntern_MultiRef(t *testing.T) {
+	mgr := &QueueManager{interner: newPool()}
```
Comment (cstyan) [inline storage/remote/intern_test.go:40]: you can just do `interner := newPool()` in each test

Comment (roidelapluie) [reply storage/remote/intern_test.go:40]: do you mean interner = newPool()? Should'nt we avoid global variables?

**Comment by roidelapluie** [comment]
And if that is really an issue, we should keep the vendors directories in our projects as it prevent users to have them in their cache.

File: storage/remote/intern_test.go (hunk 3 above)
Comment (cstyan) [reply storage/remote/intern_test.go:40]: No, `interner = newPool()` would be reusing an existing var named `interner` from somewhere. The point is that we don't need to create an instance of the `QueueManager` struct in these tests.

Comment (Harkishen-Singh) [reply storage/remote/intern_test.go]: Done!

**Comment by Harkishen-Singh** [comment]
@cstyan does that look satisfactory?

File: storage/remote/queue_manager.go [hunk 4]
```diff
@@ -303,7 +306,9 @@ func NewQueueManager(
 		samplesOut:         newEWMARate(ewmaWeight, shardUpdateDuration),
 		samplesOutDuration: newEWMARate(ewmaWeight, shardUpdateDuration),
 
-		metrics: metrics,
+		metrics:              metrics,
+		interner:             newPool(),
```
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:310]: Creating a new interner for each queue_manager will significantly increase the amount of memory used by remote write when more than one destination is configured. If we want to stop using a global interner, it should be created in https://github.com/prometheus/prometheus/blob/master/storage/remote/write.go#L64 most likely.

File: storage/remote/queue_manager_test.go [hunk 5]
```diff
@@ -47,6 +47,17 @@ import (
 
 const defaultFlushDeadline = 1 * time.Minute
 
+func getHighestTimestampGaugeMetric() *maxGauge {
```
Comment (csmarchbanks) [inline storage/remote/queue_manager_test.go:50]: This should be something like `newHighestTimestampMetric()`, `get` indicates that you are getting an existing metric which is incorrect.

**Review by csmarchbanks — CHANGES_REQUESTED** [review]
Thanks for the PR, one major concern about memory usage in the current implementation that should be easy to fix.

**Comment by Harkishen-Singh** [comment]
done @cstyan

**Review by cstyan — APPROVED** [review]
LGTM but I'll leave the merge to Chris since he requested changes last.

**Review by csmarchbanks — APPROVED** [review]
:+1: Thanks for your work!

**Pull request merged**